### ISSUE TITLE
Defining a host with teaming nics - makehosts doesn't create the 2nd nic alias

### ISSUE THREAD

**Issue by dletai** [xcat2/xcat-core#3676]
As per the [mac](http://xcat-docs.readthedocs.io/en/2.13.6/guides/admin-guides/references/man5/mac.5.html) and [hosts](http://xcat-docs.readthedocs.io/en/2.13.6/guides/admin-guides/references/man5/hosts.5.html) man pages 
1. from man mac:

> If there are multiple nics connected to Management Network(usually for bond), in order to make sure the OS deployment finished successfully, the macs of those nics must be able to resolve to same IP address. First, users have to create alias of the node for each mac in the Management Network through either: 1. adding the alias into /etc/hosts for the node directly or: 2. **setting the alias to the “hostnames” attribute** and then run “makehost” against the node.

2. from man hosts:

> Hostname aliases added to /etc/hosts for this node. Comma or blank separated list.

`chtab node=nodes hosts.hostnames='|(.*)|($1)-eno4|'`

`lsdef -i hostnames` produces a correct result, however running makehosts doesn't create this alias (while trying to debug xCAT_plugin/hosts.pm it seems the correct value is read, but is later overridden by values from nics table,

adding the alias by hand to /etc/hosts works - makedhcp creates correct entries.

It is unclear if multiple nics in same network (for bonding/teaming) should be defined in hosts or nics tables, or something else.
A clear documentation regarding defining network bonds would be appreciated.

**Comment by immarvin** [comment]
hi @bybai , would you please take a look at this? thx

**Comment by bybai** [comment]
Hi @dletai , Could you look at this documention ? http://xcat-docs.readthedocs.io/en/latest/guides/admin-guides/manage_clusters/ppc64le/diskful/customize_image/network/cfg_network_adapter.html
Here is way to create bond and aliases, I hope this doc can help you, if you have problem, contact me.Thanks.

**Comment by bybai** [comment]
I have another example to how to configure aliases and otherinterfaces for nodes in xCAT mailing list, I pasted it here:

The example:
node name in xCAT DB is node1;
Management/Deploy interface eth0, it can have alias node1-private, name in /etc/hosts and DNS is node1 and node1-private, its ip is 10.5.106.9, its network is 10_0_0_0-255_0_0_0
Secondary/Public 10Gb interface eth1, it can have alias node1-public, name in /etc/hosts and DNS node1-eth1 and node1-public,  its ip is 30.5.106.9, its network is 30_0_0_0-255_0_0_0

How to configure in xCAT DB:

1. configure network table:
]# tabdump networks
#netname,net,mask,mgtifname,gateway,dhcpserver,tftpserver,nameservers,ntpservers,logservers,dynamicrange,staticrange,staticrangeincrement,nodehostname,ddnsdomain,vlanid,domain,mtu,comments,disable
"10_0_0_0-255_0_0_0","10.0.0.0","255.0.0.0",,,,,,,,,,,,,,,,,
"30_0_0_0-255_0_0_0","30.0.0.0","255.0.0.0",,,,,,,,,,,,,,,,,

2. configure nics table:
]# lsdef node1 |sed 's/    //g'|grep ^nic
nicaliases.eth1=node1-public
nichostnamesuffixes.eth1=-eth1
nicips.eth1=30.5.106.9
nicnetworks.eth1=30_0_0_0-255_0_0_0
nictypes.eth1=ethernet

3. configure hosts table:
]# tabdump hosts
#node,ip,hostnames,otherinterfaces,comments,disable
"node1","10.5.106.9","node1-private",,,

4. create /etc/hosts
]# makehosts node1
]# cat /etc/hosts|grep 106.9
10.5.106.9 node1 node1.cluster.com node1-private
30.5.106.9 node1-eth1 node1-eth1.cluster.com node1-public

5. enable DNS
]#makedns node1
]# nslookup node1-private
Server:         10.5.106.2
Address:        10.5.106.2#53

node1-private.cluster.com    canonical name = node1.cluster.com.
Name:   node1.cluster.com
Address: 10.5.106.9

... ...

6. configure NICs in node1
if the node is provisioned, use updatenode;
]#updatenode node1 "confignetwork -s"

If the node is not provisioned, use postscript or postbootscript to call confignetwork script;

**Comment by dletai** [comment]
I have looked at the documentation page and confignetwork source before opening this issue.
confignetwork only creates bonds after the host is setup, not during anaconda, and thus doesn't care if dhcp doesn't create the secondary alias.
I require the bond to exist during installation, and have modified the template to that effect, however I had to add the alias to /etc/hosts by hand as makedhcp would fail since makehosts wouldn't create it.

makedhcp needs to know both macs, as the installation might happen over either nic, and might failover during the installation (this is rare but might happen).

In any event confingnetwork disables NetworkManager and uses the old network daemon. I would like to use teaming, not bonds, so nm is preferred, and kickstart does support it's creation (this is already working, as long as I customize the template by hand - modifying kickstartnetwork() in Template.pm to get all nics information from tables is beyond me at the moment).

**Comment by dletai** [comment]
@bybai  - the wording is from the man page for the macs table, not mine. Yes, both macs will be slaves of same master, but initially should resolve to same IP to allow the node to install over either one.

**Comment by bybai** [comment]
Hi @dletai , From your information "while trying to debug xCAT_plugin/hosts.pm it seems the correct value is read, but is later overridden by values from nics table", Could you provide the related values in nics table? Thanks.

**Comment by bybai** [comment]
@dletai , I looked at hosts.pm,  the `hostnames` in `hosts` table can be overwritten by `nicaliases` from nics table. So  could you add all your aliases into hosts table and empty the `nicalises` for this nic?
From `makehosts` initial usage:
In the `hosts` table, the `hostnames` can be added aliases for installnic.
In `nics` table, we usually consider the non-installnic configuration.
If define installnic into `nics` table, there are 2 copies aliases, one is in `hosts` table the other is in `nics` table, aliases will be overwriten. 
And `nicaliases` does not support  regex now.

**Comment by dletai** [comment]
@bybai I don't have nicaliases.
My nics line for this group looks like:

`"nodes","|\D+(\d+)|ib0!10.10.0.(10+$1),bond1!10.0.1.(10+$1),eno2!192.168.0.($1+0),eno1!10.3.1.($1+10)|","ib0!-ib0,eno1!-grg",,"eno4!Ethernet,bond1!bond,ib0!Infiniband,eno2!Ethernet,eno3!Ethernet",,,,"bond1!mode=active-backup;miimon=100","bond1!eno3|eno4",,,
`

So - no nicaliases.
makehosts creates the alias 'nodeX-bond1'. for the main (bond) IP.
This is somewhat strange as -bond1 is not in the nichostnamesufixes column.

The hosts table has this line for the group nodes:

`"nodes","|\D+(\d+)|10.0.1.(10+$1)|","|(\D+)(\d+)|($1)($2)-eno4|","|(\D+)(\d+)|($1)($2)-imm:10.2.1.($2+10)",($1)($2)-grg:10.3.1.($2+10),($1)($2)-hb:192.168.0.($2+0)|",,`

again - no hostname with suffix -bond1.
So why is -bond1 created in /etc/hosts, and why doesn't -eno4 created?

makehosts node1 ; grep node1 /etc/hosts

10.0.1.11 node1 node1.cluster node1-bond1
10.2.1.11 node1-imm node1-imm.cluster
10.3.1.11 node1-grg node1-grg.cluster
10.10.0.11 node1-ib0 node1-ib0.cluster
192.168.0.1 node1-hb node1-hb.cluster node1-eno2

**Comment by bybai** [comment]
@dletai , Could you give me your `lsxcatd -v` output?  
I used your example in my xcat MN, I does not use regex to make things easier look, I may use different ip to avoid ip conflicts in my env. I got different results with your outputs as following. For your question, why there is bond1? if there is nicips configured for bond1 and no nichostnamesuffixes configured, `makehosts` will add <nodename>-<nicname> for this nicips. The following is my example:
```
[root@bybc0602 ~]# lsdef node1
Object name: node1
    arch=x86_64
    groups=all
    hostnames=node1-eno4
    ip=10.5.106.101
    nicdevices.bond1=eno3|eno4
    nicextraparams.bond1=mode=active-backup;miimon=100
    nichostnamesuffixes.ib0=-ib0
    nichostnamesuffixes.eno1=-grg
    nicips.ib0=10.100.0.101
    nicips.bond1=10.20.1.101
    nicips.eno2=192.168.0.10
    nicips.eno1=10.3.1.110
    nictypes.eno4=Ethernet
    nictypes.bond1=bond
    nictypes.ib0=Infiniband
    nictypes.eno2=Ethernet
    nictypes.eno3=Ethernet
    otherinterfaces=node1-imm:10.2.1.11,node1-grg:10.3.1.11,node1-hb:192.168.0.1
    postbootscripts=otherpkgs
    postscripts=syslog,remoteshell,syncfiles

[root@bybc0602 ~]# tabdump nics|grep node1
"node1","ib0!10.100.0.101,bond1!10.20.1.101,eno2!192.168.0.10,eno1!10.3.1.110","ib0!-ib0,eno1!-grg",,"eno4!Ethernet,bond1!bond,ib0!Infiniband,eno2!Ethernet,eno3!Ethernet",,,,"bond1!mode=active-backup;miimon=100","bond1!eno3|eno4",,,

[root@bybc0602 ~]# tabdump hosts|grep node1
"node1","10.5.106.101","node1-eno4","node1-imm:10.2.1.11,node1-grg:10.3.1.11,node1-hb:192.168.0.1",,

[root@bybc0602 ~]# cat /etc/hosts|grep node1
10.5.106.101 node1 node1.cluster.com node1-eno4
10.2.1.11 node1-imm node1-imm.cluster.com
10.3.1.110 node1-grg node1-grg.cluster.com
192.168.0.1 node1-hb node1-hb.cluster.com
10.100.0.101 node1-ib0 node1-ib0.cluster.com
192.168.0.10 node1-eno2 node1-eno2.cluster.com
10.20.1.101 node1-bond1 node1-bond1.cluster.com

[root@bybc0602 ~]# lsxcatd -v
Version 2.13.7 (git commit 6f6a68bc37be8ea629a5e21bbb99b05f474dca51, built Thu Sep  7 06:15:37 EDT 2017)
```

**Comment by dletai** [comment]
My production is 2.13.1
    I'll try to upgrade to 2.13.7 and see if that makes a difference,
    
    Thanks for your help
    
    On 22/09//2017 06:07, Yuan Bai wrote:
    
    
      @dletai , Could you give me your lsxcatd
          -v output?
        I used your example in my xcat MN, I does not use regex to make
        things easier look, I may use different ip to avoid ip conflicts
        in my env. I got different results with your outputs as
        following. For your question, why there is bond1? if there is
        nicips configured for bond1 and no nichostnamesuffixes
        configured, makehosts will add - for this nicips.
        The following is my example:
      [root@bybc0602 ~]# lsdef node1
Object name: node1
    arch=x86_64
    groups=all
    hostnames=node1-eno4
    ip=10.5.106.101
    nicdevices.bond1=eno3|eno4
    nicextraparams.bond1=mode=active-backup;miimon=100
    nichostnamesuffixes.ib0=-ib0
    nichostnamesuffixes.eno1=-grg
    nicips.ib0=10.100.0.101
    nicips.bond1=10.20.1.101
    nicips.eno2=192.168.0.10
    nicips.eno1=10.3.1.110
    nictypes.eno4=Ethernet
    nictypes.bond1=bond
    nictypes.ib0=Infiniband
    nictypes.eno2=Ethernet
    nictypes.eno3=Ethernet
    otherinterfaces=node1-imm:10.2.1.11,node1-grg:10.3.1.11,node1-hb:192.168.0.1
    postbootscripts=otherpkgs
    postscripts=syslog,remoteshell,syncfiles

[root@bybc0602 ~]# tabdump nics|grep node1
"node1","ib0!10.100.0.101,bond1!10.20.1.101,eno2!192.168.0.10,eno1!10.3.1.110","ib0!-ib0,eno1!-grg",,"eno4!Ethernet,bond1!bond,ib0!Infiniband,eno2!Ethernet,eno3!Ethernet",,,,"bond1!mode=active-backup;miimon=100","bond1!eno3|eno4",,,

[root@bybc0602 ~]# tabdump hosts|grep node1
"node1","10.5.106.101","node1-eno4","node1-imm:10.2.1.11,node1-grg:10.3.1.11,node1-hb:192.168.0.1",,

[root@bybc0602 ~]# cat /etc/hosts|grep node1
10.5.106.101 node1 node1.cluster.com node1-eno4
10.2.1.11 node1-imm node1-imm.cluster.com
10.3.1.110 node1-grg node1-grg.cluster.com
192.168.0.1 node1-hb node1-hb.cluster.com
10.100.0.101 node1-ib0 node1-ib0.cluster.com
192.168.0.10 node1-eno2 node1-eno2.cluster.com
10.20.1.101 node1-bond1 node1-bond1.cluster.com

[root@bybc0602 ~]# lsxcatd -v
Version 2.13.7 (git commit 6f6a68bc37be8ea629a5e21bbb99b05f474dca51, built Thu Sep  7 06:15:37 EDT 2017)

      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view it on GitHub, or mute the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/xcat2/xcat-core","title":"xcat2/xcat-core","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/xcat2/xcat-core"}},"updates":{"snippets":[{"icon":"PERSON","message":"@bybai in #3676: @dletai , Could you give me your `lsxcatd -v` output?  \r\nI used your example in my xcat MN, I does not use regex to make things easier look, I may use different ip to avoid ip conflicts in my env. I got different results with your outputs as following. For your question, why there is bond1? if there is nicips configured for bond1 and no nichostnamesuffixes configured, `makehosts` will add \u003cnodename\u003e-\u003cnicname\u003e for this nicips. The following is my example:\r\n```\r\n[root@bybc0602 ~]# lsdef node1\r\nObject name: node1\r\n    arch=x86_64\r\n    groups=all\r\n    hostnames=node1-eno4\r\n    ip=10.5.106.101\r\n    nicdevices.bond1=eno3|eno4\r\n    nicextraparams.bond1=mode=active-backup;miimon=100\r\n    nichostnamesuffixes.ib0=-ib0\r\n    nichostnamesuffixes.eno1=-grg\r\n    nicips.ib0=10.100.0.101\r\n    nicips.bond1=10.20.1.101\r\n    nicips.eno2=192.168.0.10\r\n    nicips.eno1=10.3.1.110\r\n    nictypes.eno4=Ethernet\r\n    nictypes.bond1=bond\r\n    nictypes.ib0=Infiniband\r\n    nictypes.eno2=Ethernet\r\n    nictypes.eno3=Ethernet\r\n    otherinterfaces=node1-imm:10.2.1.11,node1-grg:10.3.1.11,node1-hb:192.168.0.1\r\n    postbootscripts=otherpkgs\r\n    postscripts=syslog,remoteshell,syncfiles\r\n\r\n[root@bybc0602 ~]# tabdump nics|grep node1\r\n\"node1\",\"ib0!10.100.0.101,bond1!10.20.1.101,eno2!192.168.0.10,eno1!10.3.1.110\",\"ib0!-ib0,eno1!-grg\",,\"eno4!Ethernet,bond1!bond,ib0!Infiniband,eno2!Ethernet,eno3!Ethernet\",,,,\"bond1!mode=active-backup;miimon=100\",\"bond1!eno3|eno4\",,,\r\n\r\n[root@bybc0602 ~]# tabdump hosts|grep node1\r\n\"node1\",\"10.5.106.101\",\"node1-eno4\",\"node1-imm:10.2.1.11,node1-grg:10.3.1.11,node1-hb:192.168.0.1\",,\r\n\r\n[root@bybc0602 ~]# cat /etc/hosts|grep node1\r\n10.5.106.101 node1 node1.cluster.com node1-eno4\r\n10.2.1.11 node1-imm node1-imm.cluster.com\r\n10.3.1.110 node1-grg node1-grg.cluster.com\r\n192.168.0.1 node1-hb node1-hb.cluster.com\r\n10.100.0.101 node1-ib0 node1-ib0.cluster.com\r\n192.168.0.10 node1-eno2 node1-eno2.cluster.com\r\n10.20.1.101 node1-bond1 node1-bond1.cluster.com\r\n\r\n[root@bybc0602 ~]# lsxcatd -v\r\nVersion 2.13.7 (git commit 6f6a68bc37be8ea629a5e21bbb99b05f474dca51, built Thu Sep  7 06:15:37 EDT 2017)\r\n```"}],"action":{"name":"View Issue","url":"https://github.com/xcat2/xcat-core/issues/3676#issuecomment-331338301"}}}

**Comment by bybai** [comment]
@dletai , if you have any further progress, please paste your concern here directly.

**Comment by bybai** [comment]
@dletai , since this issue has gone in 2.13.7, after you have chance to upgrade your xcat, you can retry it, now I will close this issue. thanks.

**Comment by dletai** [comment]
body p { margin-bottom: 0cm; margin-top: 0pt; } 
  
  
    Hi Yuan,
    
    
    The issue is indeed resolved in 2.13.7, thanks a lot.
    
    
    Is there any plan to switch from the "almost deprecated" network
      service to NetworkManager based configuration as per REDHAT
      documentation?
    
    
    For the intended deprecation of "network service" see the FUTURE
      section of https://access.redhat.com/solutions/783533
    
    
    One last question - how can I easily add kickstart template
      macros to define the network with anaconda based on values from
      the xcat tables?
    Currently I'm forced to use configeth script post-install,
      instead of letting anaconda define everything correctly during
      install.
    
    
    
    On 10/11/2017 04:51 AM, Yuan Bai wrote:
    
    
      Closed #3676.
      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view it on GitHub, or mute the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/xcat2/xcat-core","title":"xcat2/xcat-core","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/xcat2/xcat-core"}},"updates":{"snippets":[{"icon":"DESCRIPTION","message":"Closed #3676."}],"action":{"name":"View Issue","url":"https://github.com/xcat2/xcat-core/issues/3676#event-1287248616"}}}
